plutus-metatheory/test/TestNEAT.hs

```diff
@@ -3,103 +3,111 @@ module Main where
 import           Control.Monad.Except
 import           Data.Coolean
 import           Data.Either
+import           Data.List
+
 import           Language.PlutusCore
-import           Language.PlutusCore.Generators.NEAT.PropTest
+import           Language.PlutusCore.Evaluation.Machine.Ck
+import           Language.PlutusCore.Generators.NEAT.Spec
+import           Language.PlutusCore.Generators.NEAT.Type
+import           Language.PlutusCore.Lexer
 import           Language.PlutusCore.Normalize
 import           Test.Tasty
 import           Test.Tasty.HUnit
 
-import           MAlonzo.Code.Main                            (checkKindAgda, inferKindAgda, normalizeTypeAgda)
-import           MAlonzo.Code.Scoped                          (deBruijnifyK, unDeBruijnifyK)
+import           MAlonzo.Code.Main                         (checkKindAgda, checkTypeAgda, inferKindAgda, inferTypeAgda,
+                                                            normalizeTypeAgda, normalizeTypeTermAgda, runCKAgda,
+                                                            runLAgda, runTCEKCAgda, runTCEKVAgda, runTCKAgda)
+import           MAlonzo.Code.Scoped                       (deBruijnifyK, unDeBruijnifyK)
 
 import           Language.PlutusCore.DeBruijn
-import           Raw
-
+import           Raw                                       hiding (TypeError, tynames)
 
 main :: IO ()
-main = defaultMain allTests
-
-allTests :: TestTree
-allTests = testGroup "all tests"
-  [ testCaseCount "soundness" $
-      testTyProp size star prop_checkKindSound
-  , testCaseCount "normalization" $
-      testTyProp size star prop_normalizePreservesKind
-  , testCaseCount "normalizationSound" $
-      testTyProp size star prop_normalizeTypeSound
-  , testCaseCount "normalizationAgree" $
-      testTyProp size star prop_normalizeTypeSame
-  , testCaseCount "kindInferAgree" $
-      testTyProp size star prop_kindInferSame
+main = defaultMain $ allTests defaultGenOptions
+
+allTests :: GenOptions -> TestTree
+allTests genOpts = testGroup "NEAT"
+  [ bigTest "type-level"
+      genOpts
+      (Type ())
+      (packAssertion prop_Type)
+  , bigTest "term-level"
+      genOpts
+      (TyFunG (TyBuiltinG TyIntegerG) (TyBuiltinG TyIntegerG))
+      (packAssertion prop_Term)
   ]
 
-testCaseCount :: String -> IO Integer -> TestTree
-testCaseCount name act = testCaseInfo name $
-  liftM (\i -> show i ++ " types generated") act
-
+-- one type-level test to rule them all
```

Comment:
Did you notice any speed-up?

Reply:
Not really, I wasn't paying close enough attention to this when I changed it. But, I think it's much more sensible now.